pos_margin/models/pos_order_line.py

```diff
@@ -0,0 +1,32 @@
+# Copyright (C) 2017 - Today: GRAP (http://www.grap.coop)
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+import openerp.addons.decimal_precision as dp
+
+
+class PosOrderLine(models.Model):
+    _inherit = 'pos.order.line'
+
+    # Columns Section
+    margin = fields.Float(
+        'Margin', compute='_compute_multi_margin', store=True,
+        multi='multi_margin', digits=dp.get_precision('Product Price'))
+
+    purchase_price = fields.Float(
+        'Cost Price', compute='_compute_multi_margin', store=True,
+        multi='multi_margin', digits=dp.get_precision('Product Price'))
+
+    # Compute Section
+    @api.multi
+    @api.depends('product_id', 'qty', 'price_subtotal')
+    def _compute_multi_margin(self):
+        for line in self:
+            if not line.product_id:
+                line.purchase_price = 0
+                line.margin = 0
```

Comment:
``for line in self.filtered(lambda x: x.product_id):``

can replace line 25-> 29.

Reply:
You don't even need the lambda :slightly_smiling_face: :
```python
for line in self.filtered('product_id'):
```

Reply:
ow thanks !

Reply:
this is correct for this small piece of the code - but there is also an "else" condition in the original code. So i would have to create to for loops here - one filtered for all lines with product_id - one filtered with all lines without product_id...

I think its ok as it is

Reply:
I didn't notice that :)
An alternative:
```python
self.filtered(lambda x: not x.product_id).write({
    'purchase_price': 0,
    'margin': 0,
})
for line in self.filtered('product_id'):
    line.purchase_price = line.product_id.standard_price
    line.margin = line.price_subtotal - (
        line.product_id.standard_price * line.qty)
```

Reply:
Anyway, it can stay as it is :)

Reply:
I don't think so. I think setting default values is not necessary.
this part of code should work.

```
for line in self.filtered('product_id'):
    line.purchase_price = line.product_id.standard_price
    line.margin = line.price_subtotal - (
        line.product_id.standard_price * line.qty)
```

Reply:
i think setting default is necessary - else it will be None and not zero - this could cause problems in other calculations. So i do tend to keep the original code.
And - there is no real computation cost here - as the result gets stored - so no problem.